data/blog/reverse-engineering-ios-to-fix-sdk-crashes.mdx

```diff
@@ -0,0 +1,290 @@
+
+    if let cls = object_getClass(object) {
+        print("object_getClass:", NSStringFromClass(cls))
+
+        print("Superclass chain:")
+        var c: AnyClass? = cls
+        while let cc = c {
+            print(" ->", NSStringFromClass(cc))
+            c = class_getSuperclass(cc)
+        }
+    } else {
+        print("object_getClass returned nil")
+    }
+
+    print("obj is UISplitViewController? ->", object is UISplitViewController)
+    print("as? UISplitViewController ->", (object as? UISplitViewController) != nil)
+}
+```
+
+```
+// Output:
+
+Swift type(of:): UIPrintPanelViewController
+object_getClass: UIPrintPanelViewController
+
+Superclass chain:
+-> UIPrintPanelViewController
+-> UIViewController
+-> UIResponder
+-> NSObject
+
+obj is UISplitViewController? -> true
+as? UISplitViewController -> true
+```
+
+The output still didn't make much more sense, because it just showed once again that the `UIPrintPanelViewController` is actually a subclass of `UIViewController`, but the casting to `UISplitViewController` still succeeds.
+
+Diving further into understanding the Swift casting operator `as?` and having some discussions with our lovely LLM-companions, I remembered how we do casting in Objective-C, which is still used by most of the standard UI types in iOS:
+
+```objc
+NSObject *instance = ...; // Type-erased object
+if ([instance isKindOfClass:UITextField.self]) {
+    UITextField *textField = instance;
+    // Now use `textField` because it has type information at compile-time
+}
+```
+
+So what if the Swift operator `as?` is actually relying on the `isKindOfClass:` method?
+
+# Objective-C's Dynamic Runtime
+
+To understand what's happening and why our Cocoa SDK can perform a lot of auto-instrumentation using [Swizzling](https://docs.sentry.io/platforms/apple/guides/ios/configuration/swizzling/), we need to have a high-level understanding of how Objective-C works.
+
+When an app runs, Objective-C doesn't call methods by jumping directly to fixed memory addresses. Instead, every time compiled code calls a method, the runtime looks up which function to execute by using the method's **selector** (its name) on the class. Think of it as a dynamic “method lookup” system.
+
+Swizzling takes advantage of this. We replace the runtime's pointer for a specific method so that the selector now points to our own wrapper function instead of the original one. The original method implementation still exists — we just intercept the call so we can run our code first and then forward the call back to the real implementation. It's similar to [monkey-patching](https://en.wikipedia.org/wiki/Monkey_patch) in JavaScript or Python.
+
+Because this changes behavior _for the entire app_ and not just for our SDK, it must be done carefully.
+If two libraries swizzle the same method incorrectly or in the wrong order, it can lead to unexpected behavior.
+**This is why swizzling is powerful but also considered risky.**
```

Comment:
`m`: You could interpret this the wrong way, that we do risky stuff. Maybe we can rephrase this a bit that yes it's dangerous, but we're now what we're doing.
```suggestion
**This is why swizzling is powerful but also considered risky if you're not careful.**
```